File: base/wireworld/frontend/src/types/challenge.ts
import { CellUnion, ComponentUnion } from './wireworld';

export interface Meta {
    name: string;
    description: string;
    cell_descriptions: Record<string, string>;
    component_descriptions: Record<string, string>;
}

export interface Grid {
    width: number;
    height: number;
    max_steps: number;
}

export interface EditableArea {
    x: number;
    y: number;
    width: number;
    height: number;
}

export interface GridState {
    cell_grid: CellUnion[][];
    component_grid: (ComponentUnion | null)[][];
    step: number;
}

export interface ChallengeInfo {
    meta: Meta;
    grid: Grid;
    editable_area: EditableArea;
    initial_state: GridState;
}

export interface ChallengeResult {
    message: string;
    success: boolean | undefined;
    flag: string | null;
}


export interface SimulateMessage {
    type: 'simulate';
    design: CellUnion[][];
}

export interface GetInfoMessage {
    type: 'get_info';
}

export interface ResetMessage {
    type: 'reset';
}

export interface SimulationResultMessage {
    type: 'simulation_result';
    success: boolean;
    final_state: GridState;
    flag: string | null;
}

export interface WorldUpdateMessage {
    type: 'world_update';
    grid_state: GridState;
}

export interface ChallengeInfoMessage {
    type: 'challenge_info';
    meta: Meta;
    grid: Grid;
    editable_area: EditableArea;
    initial_state: GridState;
}

export interface ErrorMessage {
    type: 'error';
    message: string;
}

export type WebSocketMessageUnion = SimulateMessage | GetInfoMessage | ResetMessage;
export type WebSocketResponseUnion = SimulationResultMessage | WorldUpdateMessage | ChallengeInfoMessage | ErrorMessage;

File: base/wireworld/app/schemas/challenge.py
from typing import List, Optional, Union, Literal
from pydantic import BaseModel

from app.models.challenge import EditableArea, Grid, Meta
from app.simulator.wireworld import GridState
from app.simulator.cells import CellUnion


# Input message models
class SimulateMessage(BaseModel):
    type: Literal["simulate"]
    design: List[List[CellUnion]]


class GetInfoMessage(BaseModel):
    type: Literal["get_info"]


class ResetMessage(BaseModel):
    type: Literal["reset"]


# Output message models
class SimulationResultMessage(BaseModel):
    type: Literal["simulation_result"] = "simulation_result"
    success: bool
    final_state: GridState
    flag: Optional[str]


class WorldUpdateMessage(BaseModel):
    type: Literal["world_update"] = "world_update"
    grid_state: GridState


class ChallengeInfoMessage(BaseModel):
    type: Literal["challenge_info"] = "challenge_info"
    meta: Meta
    grid: Grid
    editable_area: EditableArea
    initial_state: GridState


class ErrorMessage(BaseModel):
    type: Literal["error"]
    message: str


# Union type for input messages
WebSocketMessageUnion = Union[SimulateMessage, GetInfoMessage, ResetMessage]
